File: sdks/local-rp/java/src/main/java/community/catalyst/linkkeys/localrp/crypto/Hex.java
package community.catalyst.linkkeys.localrp.crypto;

/** Lowercase hex encode/decode, no separators, no {@code 0x} prefix &mdash; the encoding every conformance vector uses. */
public final class Hex {
    private Hex() {}

    private static final char[] DIGITS = "0123456789abcdef".toCharArray();

    public static String encode(byte[] bytes) {
        char[] out = new char[bytes.length * 2];
        for (int i = 0; i < bytes.length; i++) {
            int v = bytes[i] & 0xff;
            out[i * 2] = DIGITS[v >>> 4];
            out[i * 2 + 1] = DIGITS[v & 0x0f];
        }
        return new String(out);
    }

    public static byte[] decode(String hex) {
        if (hex.length() % 2 != 0) {
            throw new IllegalArgumentException("odd-length hex string: " + hex);
        }
        byte[] out = new byte[hex.length() / 2];
        for (int i = 0; i < out.length; i++) {
            int hi = Character.digit(hex.charAt(i * 2), 16);
            int lo = Character.digit(hex.charAt(i * 2 + 1), 16);
            if (hi < 0 || lo < 0) {
                throw new IllegalArgumentException("invalid hex byte at index " + i + " in " + hex);
            }
            out[i] = (byte) ((hi << 4) | lo);
        }
        return out;
    }
}

File: sdks/local-rp/csharp/src/LinkKeys.LocalRp/Crypto/Hex.cs
namespace LinkKeys.LocalRp.Crypto;

/// <summary>
/// Lowercase hex encode/decode, no separators, no <c>0x</c> prefix — the encoding
/// every conformance vector uses.
/// </summary>
public static class Hex
{
    public static string Encode(byte[] bytes) => Convert.ToHexString(bytes).ToLowerInvariant();

    public static byte[] Decode(string hex)
    {
        if (hex.Length % 2 != 0)
        {
            throw new ArgumentException($"odd-length hex string: {hex}");
        }

        try
        {
            return Convert.FromHexString(hex);
        }
        catch (FormatException e)
        {
            throw new ArgumentException($"invalid hex string: {hex}", e);
        }
    }
}
